cmd: checkversions: A tool to check for upstream component updates

grahamwhaley opened the pull request:
Add a script that pulls version/url information from our versions.yaml
files and checks the upstream for newer versions.

It makes use of the Debian 'uscan' tool.

WIP: still need to:
 - filter out alpha versions upstream etc.
 - add support for git SHA based 'versions'

Signed-off-by: Graham Whaley <graham.whaley@intel.com>

grahamwhaley commented:
See also https://github.com/kata-containers/runtime/pull/1243 for the proposed updates to the versions.yaml file that are extracted and used by this script.

Here is a current example output. It needs some work, but will give you an idea. The reason the PASSED/FAILED are a bit messed up right now is I need to fix the regexps over on the other PR to unify the inclusion/exclusion of the leading 'v' for instance. And I'd really like to get the regexps to drop all alpha/beta/rc releases (if anybody wants to suggest a string to try.... )

```
Processing [assets.hypervisor.firecracker]
----------------------
PASSED: assets.hypervisor.firecracker.uscan-url
uscan: Newest version of assets.hypervisor.firecracker on remote site is 0.14.0, local version is v0.12.0
----------------------

Processing [assets.hypervisor.nemu]
----------------------
PASSED: assets.hypervisor.nemu.uscan-url
uscan: Newest version of assets.hypervisor.nemu on remote site is 2018-12-17, local version is release-2018-11-29
----------------------

Processing [assets.hypervisor.nemu-ovmf]
----------------------
PASSED: assets.hypervisor.nemu-ovmf.uscan-url
uscan: Newest version of assets.hypervisor.nemu-ovmf on remote site is 0.6, local version is 0.6
----------------------

Processing [assets.hypervisor.qemu]
----------------------
PASSED: assets.hypervisor.qemu.uscan-url
uscan: Newest version of assets.hypervisor.qemu on remote site is 3.1.0-rc5, local version is stable-2.11
----------------------

Processing [assets.kernel]
----------------------
PASSED: assets.kernel.uscan-url
uscan: Newest version of assets.kernel on remote site is 101, local version is v4.14.67
----------------------

Processing [externals.gometalinter]
----------------------
PASSED: externals.gometalinter.uscan-url
uscan: Newest version of externals.gometalinter on remote site is 3.0.0, local version is v2.0.5
----------------------

Processing [externals.kubernetes]
----------------------
FAILED: externals.kubernetes.uscan-url
uscan: Newest version of externals.kubernetes on remote site is 1.14.0-alpha.3, local version is 1.12.2-00
uscan:    => Newer package available from
      https://github.com/kubernetes/kubernetes/archive/v1.14.0-alpha.3.tar.gz
----------------------

Processing [externals.openshift]
----------------------
PASSED: externals.openshift.uscan-url
uscan: Newest version of externals.openshift on remote site is 4.0.0-alpha.0, local version is v3.10.0
----------------------

Processing [externals.runc]
----------------------
PASSED: externals.runc.uscan-url
uscan: Newest version of externals.runc on remote site is 1.0.0-rc6, local version is v1.0.0-rc5
----------------------

Processing [languages.golang]
----------------------
FAILED: languages.golang.uscan-url
uscan: Newest version of languages.golang on remote site is 1.12rc1, local version is 1.10.4
uscan:    => Newer package available from
      https://github.com/golang/go/archive/go1.12rc1.tar.gz
----------------------

Processing [specs.oci]
----------------------
PASSED: specs.oci.uscan-url
uscan: Newest version of specs.oci on remote site is 1.0.1, local version is v1.0.0-rc5
----------------------

PASSES: 9
FAILS: 2
```

jodh-intel commented on cmd/checkversions/checkversions.sh:
```diff
@@ -0,0 +1,90 @@
+	done
+
+	echo "PASSES: $passes"
+	echo "FAILS: $fails"
```
How about adding the following to allow the script to return a useful value:

```
exit "$fails"
```

grahamwhaley replied:
ack, good idea.

jodh-intel commented on cmd/checkversions/checkversions.sh:
```diff
@@ -0,0 +1,90 @@
+		if [ -n "$failure" ]; then
+			echo "----------------------"
+			echo "FAILED: $path"
+			echo "$result"
```
I'm not sure how easy this would be, but it could be useful to make the output of this script more machine-readable. You won't be surprised that YAML springs to mind, but even a `--csv` option that dumped out the following fields:

```
package-name;package-uscan-url;current-version;latest-version;pass/fail
```

With that, we could run this as part of the CI and even potentially have a job that attempts to:

- Update `versions.yaml` to the latest version reported by this script.
- Re-run the CI and see what happens.

grahamwhaley replied:
Yeah, I want to make the output nicer, and to do that I think I need to collect things up in a bash table first. joy!  I'm not sure I can easily extract the 'new version' from the uscan though - it doesn't hand it out in a particularly nice way - but, maybe some awk/grep can do that if we are smart. Let me see how nicely I can construct the data in the script, and then we can work on output formats.

grahamwhaley commented on cmd/checkversions/checkversions.sh:
```diff
@@ -0,0 +1,90 @@
+	local tmpjson=$(mktemp checkverXXX.json)
+	
+	${YQ} r -j "$1" > "$tmpjson"
+	paths=$(jq -r 'paths | join(".")' $tmpjson | grep "${uscan_phrase}$")
```
Probably. I might also need the json file to grab github SHA info though, in which case I'll cache it and add a bash cleanup function.

grahamwhaley commented on cmd/checkversions/checkversions.sh:
```diff
@@ -0,0 +1,90 @@
+# Look in the yaml file handed to us to find uscan URLs.
+# return a list of paths to the URL items.
+extract_uscan_items() {
+	vfile=$1
```
oops - good catch

grahamwhaley commented on cmd/checkversions/checkversions.sh:
```diff
@@ -0,0 +1,90 @@
+		debug "uscan OPTS is [$uscanopts]"
+
+		debug "Generating watchfile"
+		echo "version=4" > watchfile
```
Sure. I'll add one. It is a `uscan` oddity - confused me for a little bit to begin with ;-)

grahamwhaley commented:
Thanks for the feedback @jodh-intel yeah, this was a pretty early bit of code for review, just to show we could probably do this. I've cleaned it up somewhat and taken on board some of the ideas. Also fixed some of the parsing and a couple of regexps over on the versions.yaml. It is looking happier now. Oh, I also store the results in some arrays now before dumping the report, so it is now a bit more tabluar, and we could add yaml, csv, MD, json etc. if/when we need. Here is what it looks like now:

```$ ./checkversions.sh
Processing [assets.hypervisor.firecracker]
Processing [assets.hypervisor.nemu]
Processing [assets.hypervisor.nemu-ovmf]
Processing [assets.hypervisor.qemu]
Processing [assets.kernel]
Processing [externals.gometalinter]
Processing [externals.kubernetes]
Processing [externals.openshift]
Processing [externals.runc]
Processing [languages.golang]
Processing [specs.oci]
 Num:                                     Item          Current Ver         Upstream Ver
PASSES:
   0:    assets.hypervisor.nemu-ovmf.uscan-url                  0.6                  0.6

FAILURES:
   0:  assets.hypervisor.firecracker.uscan-url               0.12.0               0.14.0
   1:         assets.hypervisor.nemu.uscan-url           2018-11-29           2018-12-17
   2:         assets.hypervisor.qemu.uscan-url                 2.11            3.1.0-rc5
   3:                  assets.kernel.uscan-url              4.14.67             4.14.101
   4:         externals.gometalinter.uscan-url                2.0.5                3.0.0
   5:           externals.kubernetes.uscan-url            1.12.2-00       1.14.0-alpha.3
   6:            externals.openshift.uscan-url               3.10.0        4.0.0-alpha.0
   7:                 externals.runc.uscan-url            1.0.0-rc5            1.0.0-rc6
   8:               languages.golang.uscan-url               1.10.4              1.12rc1
   9:                      specs.oci.uscan-url            1.0.0-rc5                1.0.1

PASSES: 1
FAILS: 10
```

Let me see if I can figure out how to set up a naggy CI system for this now - and then add in the github SHA comparisons later.
I'll clean up the commit messages and drop the RFC/WIP then if nobody thinks this is bonkers ;-)

jodh-intel commented:
@grahamwhaley - nice - that's looking a lot more structured!

grahamwhaley commented:
@jodh-intel @chavafg @GabyCT - updated, fed in feedback, dropped all WIP status things.
Time for a proper review.
There is still a lot of scope for improvement here:
 - add in git SHA version support
 - add ability to accept path to version.yaml so we can utilise on tests repo as well

but, I think as it is it is useful, and I'd like to try and get the CI running it.
Let me clean and push the related versions.yaml update in the runtime repo as well.

chavafg commented on cmd/checkversions/checkversions.sh:
```diff
@@ -0,0 +1,139 @@
+	done
+}
+
+# Generate a nice tabluated summary of the results.
```
s/tabluated/tabulated/ ?

grahamwhaley replied:
heh, and I think I did it twice... fixing. will wait for any other fold ins...

grahamwhaley replied:
typos fixed and pushed..

grahamwhaley commented:
/test

GabyCT commented:
/test

chavafg commented:
/test
